src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinderProvider.cs

```diff
@@ -21,12 +24,19 @@ public IModelBinder GetBinder(ModelBinderProviderContext context)
             if (context.BindingInfo.BindingSource != null &&
                 context.BindingInfo.BindingSource.CanAcceptDataFrom(BindingSource.Header))
             {
+                var loggerFactory = context.Services.GetRequiredService<ILoggerFactory>();
+                var logger = loggerFactory.CreateLogger<HeaderModelBinderProvider>();
+
                 // We only support strings and collections of strings. Some cases can fail
                 // at runtime due to collections we can't modify.
                 if (context.Metadata.ModelType == typeof(string) ||
                     context.Metadata.ElementType == typeof(string))
                 {
-                    return new HeaderModelBinder();
+                    return new HeaderModelBinder(loggerFactory);
+                }
+                else
+                {
+                    logger.CannotCreateHeaderModelBinder(context.Metadata.ModelType);
```

Comment:
I did for this because I ran into an issue where I decorated a `int` parameter with [FromHeader] but the value was not bound even though i sent a valid(a number) value in the header. This similar scenario usually works, for example when the parameter is decorated with [FromQuery]. So, that is why I wanted to explicitly log this information to let users know that it only works with model types of string or collection of string

Reply:
Hmm, I guess this is alright. We don't have many model binder providers with multiple distinct match requirements. So, the case may actually be unique.